src/analysis/normed_space/compact_operator.lean

```diff
@@ -0,0 +1,365 @@
+
+lemma compact_operator.neg [has_continuous_neg M₄] {f : M₁ → M₄}
+  (hf : compact_operator f) : compact_operator (-f) :=
+let ⟨K, hK, hKf⟩ := hf in
+⟨-K, hK.neg, mem_of_superset hKf $ λ x (hx : f x ∈ K), set.neg_mem_neg.mpr hx⟩
+
+lemma compact_operator.sub [topological_add_group M₄] {f g : M₁ → M₄}
+  (hf : compact_operator f) (hg : compact_operator g) : compact_operator (f - g) :=
+by rw sub_eq_add_neg; exact hf.add hg.neg
+
+end operations
+
+section comp
+
+variables {R₁ R₂ R₃ : Type*} [semiring R₁] [semiring R₂] [semiring R₃] {σ₁₂ : R₁ →+* R₂}
+  {σ₂₃ : R₂ →+* R₃} {M₁ M₂ M₃ : Type*} [topological_space M₁] [topological_space M₂]
+  [topological_space M₃] [add_comm_monoid M₁] [module R₁ M₁]
+
+lemma compact_operator.comp_clm [add_comm_monoid M₂] [module R₂ M₂] {f : M₂ → M₃}
+  (hf : compact_operator f) (g : M₁ →SL[σ₁₂] M₂) :
+  compact_operator (f ∘ g) :=
+begin
+  have := g.continuous.tendsto 0,
+  rw map_zero at this,
+  rcases hf with ⟨K, hK, hKf⟩,
+  exact ⟨K, hK, this hKf⟩
+end
+
+lemma compact_operator.continuous_comp {f : M₁ → M₂} (hf : compact_operator f) {g : M₂ → M₃}
+  (hg : continuous g) :
+  compact_operator (g ∘ f) :=
+begin
+  rcases hf with ⟨K, hK, hKf⟩,
+  refine ⟨g '' K, hK.image hg, mem_of_superset hKf _⟩,
+  nth_rewrite 1 preimage_comp,
+  exact preimage_mono (subset_preimage_image _ _)
+end
+
+lemma compact_operator.clm_comp [add_comm_monoid M₂] [module R₂ M₂] [add_comm_monoid M₃]
+  [module R₃ M₃] {f : M₁ → M₂} (hf : compact_operator f) (g : M₂ →SL[σ₂₃] M₃) :
+  compact_operator (g ∘ f) :=
+hf.continuous_comp g.continuous
+
+lemma compact_operator.continuous_on_comp [t2_space M₂] {f : M₁ → M₂} (hf : compact_operator f)
+  {g : M₂ → M₃} (hg : continuous_on g (closure $ range f)) :
+  compact_operator (g ∘ f) :=
+begin
+  rw compact_operator_iff_exists_mem_nhds_image_rel_compact at hf,
+  rcases hf with ⟨V, hV, hVf⟩,
+  refine ⟨g '' (closure $ f '' V),
+          hVf.image_of_continuous_on (hg.mono $ closure_mono $ image_subset_range _ _),
+          mem_of_superset hV _⟩,
+  calc
+    V ⊆ f ⁻¹' (f '' V) : subset_preimage_image _ _
+  ... ⊆ f ⁻¹' (closure $ f '' V) : preimage_mono subset_closure
+  ... ⊆ f ⁻¹' (g ⁻¹' (g '' (closure $ f '' V))) : preimage_mono (subset_preimage_image _ _)
+end
+
+end comp
+
+section restrict_invariant
+
+variables {R₁ R₂ R₃ : Type*} [semiring R₁] [semiring R₂] [semiring R₃] {σ₁₂ : R₁ →+* R₂}
+  {σ₂₃ : R₂ →+* R₃} {M₁ M₂ M₃ : Type*} [topological_space M₁] [uniform_space M₂]
+  [topological_space M₃] [add_comm_monoid M₁] [add_comm_monoid M₂] [add_comm_monoid M₃]
+  [module R₁ M₁] [module R₂ M₂] [module R₃ M₃]
+
+/-- If a compact operator preserves a closed submodule, its restriction to that submodule is
+compact. -/
+lemma compact_operator.restrict_invariant {f: M₁ →ₗ[R₁] M₁} (hf : compact_operator f)
+  {V : submodule R₁ M₁} (hV : ∀ v ∈ V, f v ∈ V) (h_closed : is_closed (V : set M₁)):
+  compact_operator (f.restrict hV) :=
```

Comment:
To me, this lemma is naturally split into two facts:
(1) if a compact operator takes values in a closed subspace, then the restricted operator (with the same source space, but restricted target space) is compact
(2) The inclusion of a subspace in the whole space is a continuous linear map
Since the composition of a compact operator and a continuous linear map is still compact, then your lemma follows directly. 

Could you formulate and prove (1) first (using `cod_restrict`), and then deduce your lemma from it?